src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -629,81 +631,140 @@ HRESULT STDMETHODCALLTYPE CorProfiler::ModuleLoadFinished(ModuleID module_id, HR
+        CallTarget_RequestRejitForModule(module_id, module_metadata, integration_methods_);
+    }
+    else
     {
-        filtered_integrations = FilterIntegrationsByTarget(filtered_integrations, assembly_import);
+        std::vector<IntegrationMethod> filtered_integrations =
+            FilterIntegrationsByCaller(integration_methods_, module_info.assembly);
 
         if (filtered_integrations.empty())
         {
             // we don't need to instrument anything in this module, skip it
-            Logger::Debug("ModuleLoadFinished skipping module (filtered by target): ", module_id, " ",
-                          module_info.assembly.name);
+            Logger::Debug("ModuleLoadFinished skipping module (filtered by caller): ", module_id, " ", module_info.assembly.name);
             return S_OK;
         }
-    }
 
-    mdModule module;
-    hr = metadata_import->GetModuleFromScope(&module);
-    if (FAILED(hr))
-    {
-        Logger::Warn("ModuleLoadFinished failed to get module metadata token for ", module_id, " ", module_info.assembly.name);
-        return S_OK;
-    }
+        auto hr = this->info_->GetModuleMetaData(module_id, ofRead | ofWrite, IID_IMetaDataImport2, metadata_interfaces.GetAddressOf());
 
-    GUID module_version_id;
-    hr = metadata_import->GetScopeProps(nullptr, 0, nullptr, &module_version_id);
-    if (FAILED(hr))
-    {
-        Logger::Warn("ModuleLoadFinished failed to get module_version_id for ", module_id, " ", module_info.assembly.name);
-        return S_OK;
-    }
+        if (FAILED(hr))
+        {
+            Logger::Warn("ModuleLoadFinished failed to get metadata interface for ", module_id, " ", module_info.assembly.name);
+            return S_OK;
+        }
 
-    ModuleMetadata* module_metadata =
-        new ModuleMetadata(metadata_import, metadata_emit, assembly_import, assembly_emit, module_info.assembly.name,
-                           app_domain_id, module_version_id, filtered_integrations, &corAssemblyProperty);
+        const auto metadata_import = metadata_interfaces.As<IMetaDataImport2>(IID_IMetaDataImport);
+        const auto metadata_emit = metadata_interfaces.As<IMetaDataEmit2>(IID_IMetaDataEmit);
+        const auto assembly_import = metadata_interfaces.As<IMetaDataAssemblyImport>(IID_IMetaDataAssemblyImport);
+        const auto assembly_emit = metadata_interfaces.As<IMetaDataAssemblyEmit>(IID_IMetaDataAssemblyEmit);
 
-    // store module info for later lookup
-    module_id_to_info_map_[module_id] = module_metadata;
+        // don't skip Microsoft.AspNetCore.Hosting so we can run the startup hook and
+        // subscribe to DiagnosticSource events.
+        // don't skip Dapper: it makes ADO.NET calls even though it doesn't reference
+        // System.Data or System.Data.Common
+        if (module_info.assembly.name != WStr("Microsoft.AspNetCore.Hosting") && module_info.assembly.name != WStr("Dapper"))
```

Comment:
I think these strings could also be replaced with constants so they're not allocated on every module load

Reply:
done in 988f718